### ISSUE TITLE
GUI daemon occasionally does not start

### ISSUE THREAD

**Issue by DemiMarie** [QubesOS/qubes-issues#4729]
### Qubes OS version:

`Qubes release 4.0 (R4.0)`

### Affected component(s): 

dom0 GUI daemon

---

### Steps to reproduce the behavior:

Reboot the system many times

### Expected behavior:

`qubes-guid` spawns for each VM every time

### Actual behavior:

Rarely (maybe 10% of the time?), `qubes-guid` doesn’t start.  It works fine when started manually.

### General notes:

The problem goes away on reboot.

---

**Comment by marmarek** [comment]
Is `qvm-start-gui` process running? See `~/.xsession-errors` for possible related messages.

**Comment by DemiMarie** [comment]
Not sure.  I will check when the issue happens again.

**Comment by iprid23** [comment]
@marmarek Sometimes after I restart X, the qubes-guid is not working. 
Trying to manually start it like this, I get:
/usr/bin/qubes-guid -N ... -c 0xcc0000 -i /usr/share/icons/hicolor/128x128/devices/appvm-yellow.png -l 3 -d 23
```
Created 0x...(0x...) parent 0x0(0x3aa) ovr=0 x/y 10/10 w/h 10/10
  do_shm_update for 0x...(remote 0x...), after border calc: x=0, y=0, w=1000, h=974
handle_configure_from_vm, local 0x... remote 0x1000001, 10/10, was 10/10, ovr=0, xy 10/10, was 10/10
ErrorHandler: BadValue (integer parameter out of range for operation)
                 Major opcode: 130 (MIT-SHM)
                 Minor opcode: 3 (X_ShmPutImage)
                 Value:        0x3e8
                 Failed serial number:  126
                 Current serial number: 127
```

**Comment by marmarek** [comment]
@iprid23 this problem may be a separate one. Note that original report says "It works fine when started manually", while in your case it failed to start this way.
Similar issue is already fixed in master branch (https://github.com/QubesOS/qubes-gui-daemon/commit/da147e219acbe02cdac64a2cf503b028ca6fbc31), but that one is about XShmAttach, not XShmPutImage. It might help anyway though.
You've removed window IDs - can you confirm all those messages are about the same window. Did you get any other message about that window?

@jpouellet can you paste the message you've got recently? Was that X_ShmPutImage too?

**Comment by jpouellet** [comment]
```
[user@dom0 ~]$ LANG=C grep -h -r -C1000 X_ /var/log/qubes/guid.*
Icon size: 128x128
ErrorHandler: BadAccess (attempt to access private resource denied)
                 Major opcode: 130 (MIT-SHM)
                 Minor opcode: 1 (X_ShmAttach)
                 ResourceID:   0x380001c
                 Failed serial number:  304
                 Current serial number: 305
```

This happened to a DispVM after logging out and logging back in (of the dom0 xfce session).

**Comment by marmarek** [comment]
Ok, so this one is exactly what https://github.com/QubesOS/qubes-gui-daemon/commit/da147e219acbe02cdac64a2cf503b028ca6fbc31 fixes. I'm not really sure if that commit will help with XShmPutImage, but it may (by setting `vm_window->image = NULL` and avoiding that XShmPutImage call at all).

**Comment by ra--** [comment]
I'm experiencing the same issue as the OP, i.e., qubes-guid not starting occasionally. I'm actually encountering that issue for a few months now but just got so annoyed by it today that I decided to look into it a bit.
After a couple of days of uptime qubes-guid is not started anymore for new VMs, which results in graphical applications from new VMs not being available any more. As soon as I start qubes-guid manually in dom0, the expected behavior is back. Unfortunately, I cannot find anything related to qubes-guid startup in the logs (i.e., /var/log/qubes/* and ~/.xsession-errors). Any pointers where the start of qubes-guid is initialized?

**Comment by ra--** [comment]
Followup: The error doesn't seem be related to executing qubes-guid, but qubes-guid is not even started to begin with (tested by replacing /usr/bin/qubes-guid with a simple script that logs its execution). 
It seems that the start of qubes-guid is initiated in qubes-core-admin/qubes/vm/qubesvm.py and from skimming the code it looks ok, but I'm nowhere familiar with the Qubes' code base.

**Comment by marmarek** [comment]
`qubes-guid` is started by `qvm-start-gui` process, which should be running as part of your login session (not qubesd service). See if you see process `qvm-start-gui --all --watch`. If not, it probably crashed for some reason and left some clues in `~/.xsession-errors`.

**Comment by ra--** [comment]
qvm-start-gui wasn't running and (re)starting did the trick. Thank you!

Find below what seem to be the entries in ~/.xsession-errors corresponding to the crash (using qubes-core-admin-client-4.0.28-1.fc25.noarch):

app: Connection to qubesd terminated, reconnecting in 1.0 seconds
Connection to qubesd terminated, reconnecting in 1.0 seconds
app: Connection to qubesd terminated, reconnecting in 1.0 seconds
Connection to qubesd terminated, reconnecting in 1.0 seconds
Traceback (most recent call last):
  File "/usr/bin/qvm-start-gui", line 5, in <module>
    sys.exit(main())
  File "/usr/lib/python3.5/site-packages/qubesadmin/tools/qvm_start_gui.py", line 426, in main
    loop.run_until_complete(events_listener)
  File "/usr/lib64/python3.5/asyncio/base_events.py", line 467, in run_until_complete
    return future.result()
  File "/usr/lib64/python3.5/asyncio/futures.py", line 294, in result
    raise self._exception
  File "/usr/lib64/python3.5/asyncio/tasks.py", line 240, in _step
    result = coro.send(None)
  File "/usr/lib/python3.5/site-packages/qubesadmin/events/__init__.py", line 126, in listen_for_events
    yield from self._listen_for_events(vm)
  File "/usr/lib/python3.5/site-packages/qubesadmin/events/__init__.py", line 183, in _listen_for_events
    self.handle(subject, event, **kwargs)
  File "/usr/lib/python3.5/site-packages/qubesadmin/events/__init__.py", line 217, in handle
    handler(subject, event, **kwargs)
  File "/usr/lib/python3.5/site-packages/qubesadmin/tools/qvm_start_gui.py", line 341, in on_connection_established
    monitor_layout = get_monitor_layout()
  File "/usr/lib/python3.5/site-packages/qubesadmin/tools/qvm_start_gui.py", line 81, in get_monitor_layout
    output_params = REGEX_OUTPUT.match(line).groupdict()
AttributeError: 'NoneType' object has no attribute 'groupdict'
Exception ignored in: <bound method BaseEventLoop.__del__ of <_UnixSelectorEventLoop running=False closed=True debug=False>>
Traceback (most recent call last):
  File "/usr/lib64/python3.5/asyncio/base_events.py", line 511, in __del__
  File "/usr/lib64/python3.5/asyncio/unix_events.py", line 65, in close
  File "/usr/lib64/python3.5/asyncio/unix_events.py", line 146, in remove_signal_handler
  File "/usr/lib64/python3.5/signal.py", line 47, in signal
TypeError: signal handler must be signal.SIG_IGN, signal.SIG_DFL, or a callable object

**Comment by marmarek** [comment]
Do you see any message suggesting `xrandr` have failed somewhere there?

**Comment by ra--** [comment]
Not xrandr specifically, but some error message that are related to handling changes to the screen layout (before the crash - but since there is no exact time/date information in the log entries might not be closely related timewise):

// One line for each running VM:
Failed to send monitor layout: b'gntshr: error: ioctl failed: No space left on device\nFailed to start data vchan server\n'

// Many entries of successful(?) changes to the screen layout

Screen layout change event received
asyncio: Task exception was never retrieved
future: <Task finished coro=<GUILauncher.start_gui_for_vm() done, defined at /usr/lib/python3.5/site-packages/qubesadmin/tools/qvm_start_gui.py:183> exception=AttributeError("'NoneType' object has no attribute 'groupdict'",)>
Traceback (most recent call last):
  File "/usr/lib64/python3.5/asyncio/tasks.py", line 240, in _step
    result = coro.send(None)
  File "/usr/lib/python3.5/site-packages/qubesadmin/tools/qvm_start_gui.py", line 211, in start_gui_for_vm
    startup=True)
  File "/usr/lib/python3.5/site-packages/qubesadmin/tools/qvm_start_gui.py", line 278, in send_monitor_layout
    layout = get_monitor_layout()
  File "/usr/lib/python3.5/site-packages/qubesadmin/tools/qvm_start_gui.py", line 81, in get_monitor_layout
    output_params = REGEX_OUTPUT.match(line).groupdict()
AttributeError: 'NoneType' object has no attribute 'groupdict'
asyncio: Exception in callback GUILauncher.send_monitor_layout_all()
handle: <Handle GUILauncher.send_monitor_layout_all()>
Traceback (most recent call last):
  File "/usr/lib64/python3.5/asyncio/events.py", line 127, in _run
    self._callback(*self._args)
  File "/usr/lib/python3.5/site-packages/qubesadmin/tools/qvm_start_gui.py", line 306, in send_monitor_layout_all
    monitor_layout = get_monitor_layout()
  File "/usr/lib/python3.5/site-packages/qubesadmin/tools/qvm_start_gui.py", line 81, in get_monitor_layout
    output_params = REGEX_OUTPUT.match(line).groupdict()
AttributeError: 'NoneType' object has no attribute 'groupdict'

**Comment by ra--** [comment]
I might have identified the root cause for the crash now:
After some time, I get an additional "ghost screen":
```
$ xrandr -q
..
DP2-1 unknown connection (normal left inverted right x axis y axis)
..
```
The "unknown connection" part is not covered by REGEX_OUTPUT on line 54 qvm_start_gui.py, which is the cause for the crash I think.

**Comment by csarn** [comment]
I have the same issue (gui daemon not starting due to "unknown connection" not being handled).
But it is not "occasionally". I get this every time that I boot up my Thinkpad W530 (with only NVidia graphics enabled) in my docking station. When booting outside of the dock, LVDS-1 is connected, but when docked, LVDS-1 shows "unknown connection" in xrandr.

This is in qubes 4.1 alpha.

**Comment by donob4n** [comment]
I am not sure if is related to this but I had a strange GUI crash when logging into KDE.
[gui-crash.txt](https://github.com/QubesOS/qubes-issues/files/5980719/gui-crash.txt)

**Comment by DemiMarie** [comment]
> ```
> [user@dom0 ~]$ LANG=C grep -h -r -C1000 X_ /var/log/qubes/guid.*
> Icon size: 128x128
> ErrorHandler: BadAccess (attempt to access private resource denied)
>                  Major opcode: 130 (MIT-SHM)
>                  Minor opcode: 1 (X_ShmAttach)
>                  ResourceID:   0x380001c
>                  Failed serial number:  304
>                  Current serial number: 305
> ```
> 
> This happened to a DispVM after logging out and logging back in (of the dom0 xfce session).

There appears to be a GUI agent bug that can cause it to pass invalid grant requests to the GUI daemon.  I have not figured out why it does that, though.

**Comment by the-moog** [comment]
I have a similar issue after switching to KDE.  How and when (and by which user) is qvm-start-gui supposed to be started?

I could not find any systemd entries for it so I put a KDE startup item in, though it works, that seems wrong to me.

I can see that it affects auto-started VMs as these start before the user logs in (or take a shile to start) and if this service is missing those auto starts can fail, e.g. USB does this a lot.

**Comment by marmarek** [comment]
> How and when (and by which user) is qvm-start-gui supposed to be started?

It's started via `/etc/xdg/autostart/qvm-start-daemon.desktop` (or -kde variant under KDE).

**Comment by the-moog** [comment]
Ok in that case there is some sort of install bug here as the file(s) qvm-start-daemon(-kde).desktop are entirely missing after switching to KDE.  This makes me wonder what else did not install properly.  
I replace from https://github.com/QubesOS/qubes-core-admin-client/tree/main/etc but is there a reason why this may have happened?

**Comment by the-moog** [comment]
Being a .desktop file means that it's run after the desktop environment is spawned.  Any comment on the possible race condition with system VMs (e.g. net/usb) as a result?

**Comment by marmarek** [comment]
> Being a .desktop file means that it's run after the desktop environment is spawned

Yes, that's exactly why it's done this way - GUI daemon must start only after user logs in, to be able to show application windows.

> Any comment on the possible race condition with system VMs (e.g. net/usb) as a result?

qvm-start-daemon normally handles also already-running VMs. If there is some issue with that, you should get a message about it in `~/.xsession-errors`. There were some issues with that, but last one I'm aware of was fixed in April.

**Comment by the-moog** [comment]
Yes. I do get messages about it in .xsession-errors. The usb qube tries to start before the user's gui starts at login. It then fails (rather than waits) because it has no gui.  Hence the mouse does not work until you manually start it.  Sometimes the net qube does the same (for for some reason less often).
Perhaps qubes should have a 'needs' field.  So that they will suspend start until their dependency is met, spawning that dependency as necessary?

**Comment by marmarek** [comment]
> The usb qube tries to start before the user's gui starts at login. It then fails (rather than waits) because it has no gui.

USB qube (or net qube for that matter) shouldn't need GUI to function, including proxying USB mouse. Can you post exact error message you get when it "fails"?

**Comment by the-moog** [comment]
I can only see this after a boot.  Manual starts overwrite the logs.
I am running build jobs right now but on the next reboot I can get this info.

**Comment by andrewdavidwong** [comment]
Is this still a problem in 4.1?

**Comment by ra--** [comment]
I haven't experienced that issue in a long time. Not sure what has fixed it, because the regex mentioned in https://github.com/QubesOS/qubes-issues/issues/4729#issuecomment-596422187 doesn't seem to have been updated. :man_shrugging:

**Comment by andrewdavidwong** [comment]
> I haven't experienced that issue in a long time. Not sure what has fixed it, because the regex mentioned in [#4729 (comment)](https://github.com/QubesOS/qubes-issues/issues/4729#issuecomment-596422187) doesn't seem to have been updated. man_shrugging

Closing. If anyone is still affected by this issue, please leave a comment, and we'll be happy to reopen this. Thank you.
